File: src/C3.Infrastructure/CatalogueFiles/Xml/V1_1/LegacyCassetteModelRepository.vb
Imports C3.Catalogue.CassetteModels
Imports System.Data

Namespace CatalogueFiles.Xml.V1_1

    Public NotInheritable Class LegacyCassetteModelRepository
        Implements ICassetteModelRepository

        Private ReadOnly _documentProvider As Func(Of DataSet)

        Public Sub New(documentProvider As Func(Of DataSet))
            If documentProvider Is Nothing Then
                Throw New ArgumentNullException("documentProvider")
            End If
            _documentProvider = documentProvider
        End Sub

        Public Function GetAll() As IList(Of CassetteModel) Implements ICassetteModelRepository.GetAll
            Dim values As New List(Of CassetteModel)()
            For Each row As DataRow In ModelsTable().Rows
                If row.RowState <> DataRowState.Deleted Then
                    values.Add(Map(row))
                End If
            Next
            Return values
        End Function

        Public Function FindByIdentifier(identifier As String) As CassetteModel _
                Implements ICassetteModelRepository.FindByIdentifier

            Dim row As DataRow = FindRow(identifier)
            If row Is Nothing Then
                Return Nothing
            End If
            Return Map(row)
        End Function

        Public Function BrandExists(code As String) As Boolean _
                Implements ICassetteModelRepository.BrandExists

            For Each row As DataRow In RequireTable("Brands").Rows
                If row.RowState <> DataRowState.Deleted AndAlso
                        String.Equals(Convert.ToString(row("Code")), code, StringComparison.OrdinalIgnoreCase) Then
                    Return True
                End If
            Next
            Return False
        End Function

        Public Function IdentifierExists(identifier As String) As Boolean _
                Implements ICassetteModelRepository.IdentifierExists

            Return FindRow(identifier) IsNot Nothing
        End Function

        Public Function IsReferencedByTape(identifier As String) As Boolean _
                Implements ICassetteModelRepository.IsReferencedByTape

            For Each row As DataRow In RequireTable("Tapes").Rows
                If row.RowState <> DataRowState.Deleted AndAlso
                        String.Equals(Convert.ToString(row("Model")), identifier, StringComparison.OrdinalIgnoreCase) Then
                    Return True
                End If
            Next
            Return False
        End Function

        Public Sub Add(value As CassetteModel) Implements ICassetteModelRepository.Add
            ModelsTable().Rows.Add(
                value.BrandCode,
                value.TypeNumber,
                value.ModelName,
                value.Code,
                value.Identifier,
                value.DisplayName,
                value.TapeCount,
                value.AddedAt,
                value.Notes)
            SynchronizeModelCounter()
        End Sub

        Public Sub Update(value As CassetteModel) Implements ICassetteModelRepository.Update
            Dim row As DataRow = FindRow(value.Identifier)
            If row Is Nothing Then
                Throw New InvalidOperationException("The selected cassette model no longer exists.")
            End If

            row("Model") = value.ModelName
            row("Name") = value.DisplayName
            row("Notes") = value.Notes
        End Sub

        Public Sub Delete(identifier As String) Implements ICassetteModelRepository.Delete
            Dim row As DataRow = FindRow(identifier)
            If row Is Nothing Then
                Throw New InvalidOperationException("The selected cassette model no longer exists.")
            End If

            ModelsTable().Rows.Remove(row)
            SynchronizeModelCounter()
        End Sub

        Private Function Document() As DataSet
            Dim value As DataSet = _documentProvider()
            If value Is Nothing Then
                Throw New InvalidOperationException("No active catalogue document is available.")
            End If
            Return value
        End Function

        Private Function ModelsTable() As DataTable
            Return RequireTable("Models")
        End Function

        Private Function RequireTable(name As String) As DataTable
            Dim table As DataTable = Document().Tables(name)
            If table Is Nothing Then
                Throw New InvalidOperationException("Catalogue table '" & name & "' is missing.")
            End If
            Return table
        End Function

        Private Function FindRow(identifier As String) As DataRow
            If String.IsNullOrWhiteSpace(identifier) Then
                Return Nothing
            End If

            For Each row As DataRow In ModelsTable().Rows
                If row.RowState <> DataRowState.Deleted AndAlso
                        String.Equals(
                            Convert.ToString(row("Identifier")),
                            identifier,
                            StringComparison.OrdinalIgnoreCase) Then
                    Return row
                End If
            Next
            Return Nothing
        End Function

        Private Function Map(row As DataRow) As CassetteModel
            Dim addedAt As DateTime = DateTime.MinValue
            If Not row.IsNull("Date") Then
                addedAt = Convert.ToDateTime(row("Date"))
            End If

            Return New CassetteModel(
                ResolveBrandCode(Convert.ToString(row("Brand"))),
                ReadInteger(row, "Type"),
                Convert.ToString(row("Model")),
                Convert.ToString(row("Code")),
                Convert.ToString(row("Identifier")),
                Convert.ToString(row("Name")),
                ReadInteger(row, "Number"),
                addedAt,
                Convert.ToString(row("Notes")))
        End Function

        Private Function ResolveBrandCode(storedValue As String) As String
            For Each row As DataRow In RequireTable("Brands").Rows
                If row.RowState <> DataRowState.Deleted Then
                    Dim code As String = Convert.ToString(row("Code"))
                    Dim name As String = Convert.ToString(row("Brand"))
                    If String.Equals(storedValue, code, StringComparison.OrdinalIgnoreCase) OrElse
                            String.Equals(storedValue, name, StringComparison.OrdinalIgnoreCase) Then
                        Return code
                    End If
                End If
            Next
            Return storedValue
        End Function

        Private Shared Function ReadInteger(row As DataRow, columnName As String) As Integer
            If row.IsNull(columnName) Then
                Return 0
            End If
            Return Convert.ToInt32(row(columnName))
        End Function

        Private Sub SynchronizeModelCounter()
            Dim counter As DataRow = RequireTable("Counters").Rows.Find("Models")
            If counter IsNot Nothing Then
                counter("Number") = ModelsTable().Rows.Count
            End If
        End Sub

    End Class

End Namespace

File: src/C3.Infrastructure/CatalogueFiles/Xml/V1_1/LegacyCassetteModelRepository.cs
using C3.Catalogue.CassetteModels;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Runtime.CompilerServices;

namespace C3.Infrastructure.CatalogueFiles.Xml.V1_1
{
	public sealed class LegacyCassetteModelRepository : ICassetteModelRepository
	{
		private readonly Func<DataSet> _documentProvider;

		public LegacyCassetteModelRepository(Func<DataSet> documentProvider)
		{
			if (documentProvider == null)
			{
				throw new ArgumentNullException("documentProvider");
			}
			_documentProvider = documentProvider;
		}

		public IList<CassetteModel> GetAll()
		{
			List<CassetteModel> list = new List<CassetteModel>();
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = ModelsTable().Rows.GetEnumerator();
				while (enumerator.MoveNext())
				{
					DataRow dataRow = (DataRow)enumerator.Current;
					if (dataRow.RowState != DataRowState.Deleted)
					{
						list.Add(Map(dataRow));
					}
				}
				return list;
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
		}

		IList<CassetteModel> ICassetteModelRepository.GetAll()
		{
			//ILSpy generated this explicit interface implementation from .override directive in GetAll
			return this.GetAll();
		}

		public CassetteModel FindByIdentifier(string identifier)
		{
			DataRow dataRow = FindRow(identifier);
			if (dataRow == null)
			{
				return null;
			}
			return Map(dataRow);
		}

		CassetteModel ICassetteModelRepository.FindByIdentifier(string identifier)
		{
			//ILSpy generated this explicit interface implementation from .override directive in FindByIdentifier
			return this.FindByIdentifier(identifier);
		}

		public bool BrandExists(string code)
		{
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = RequireTable("Brands").Rows.GetEnumerator();
				while (enumerator.MoveNext())
				{
					DataRow dataRow = (DataRow)enumerator.Current;
					if (dataRow.RowState != DataRowState.Deleted && string.Equals(Convert.ToString(RuntimeHelpers.GetObjectValue(dataRow["Code"])), code, StringComparison.OrdinalIgnoreCase))
					{
						return true;
					}
				}
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
			return false;
		}

		bool ICassetteModelRepository.BrandExists(string code)
		{
			//ILSpy generated this explicit interface implementation from .override directive in BrandExists
			return this.BrandExists(code);
		}

		public bool IdentifierExists(string identifier)
		{
			return FindRow(identifier) != null;
		}

		bool ICassetteModelRepository.IdentifierExists(string identifier)
		{
			//ILSpy generated this explicit interface implementation from .override directive in IdentifierExists
			return this.IdentifierExists(identifier);
		}

		public bool IsReferencedByTape(string identifier)
		{
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = RequireTable("Tapes").Rows.GetEnumerator();
				while (enumerator.MoveNext())
				{
					DataRow dataRow = (DataRow)enumerator.Current;
					if (dataRow.RowState != DataRowState.Deleted && string.Equals(Convert.ToString(RuntimeHelpers.GetObjectValue(dataRow["Model"])), identifier, StringComparison.OrdinalIgnoreCase))
					{
						return true;
					}
				}
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
			return false;
		}

		bool ICassetteModelRepository.IsReferencedByTape(string identifier)
		{
			//ILSpy generated this explicit interface implementation from .override directive in IsReferencedByTape
			return this.IsReferencedByTape(identifier);
		}

		public void Add(CassetteModel value)
		{
			ModelsTable().Rows.Add(value.BrandCode, value.TypeNumber, value.ModelName, value.Code, value.Identifier, value.DisplayName, value.TapeCount, value.AddedAt, value.Notes);
			SynchronizeModelCounter();
		}

		void ICassetteModelRepository.Add(CassetteModel value)
		{
			//ILSpy generated this explicit interface implementation from .override directive in Add
			this.Add(value);
		}

		public void Update(CassetteModel value)
		{
			DataRow dataRow = FindRow(value.Identifier);
			if (dataRow == null)
			{
				throw new InvalidOperationException("The selected cassette model no longer exists.");
			}
			dataRow["Model"] = value.ModelName;
			dataRow["Name"] = value.DisplayName;
			dataRow["Notes"] = value.Notes;
		}

		void ICassetteModelRepository.Update(CassetteModel value)
		{
			//ILSpy generated this explicit interface implementation from .override directive in Update
			this.Update(value);
		}

		public void Delete(string identifier)
		{
			DataRow dataRow = FindRow(identifier);
			if (dataRow == null)
			{
				throw new InvalidOperationException("The selected cassette model no longer exists.");
			}
			ModelsTable().Rows.Remove(dataRow);
			SynchronizeModelCounter();
		}

		void ICassetteModelRepository.Delete(string identifier)
		{
			//ILSpy generated this explicit interface implementation from .override directive in Delete
			this.Delete(identifier);
		}

		private DataSet Document()
		{
			DataSet dataSet = _documentProvider();
			if (dataSet == null)
			{
				throw new InvalidOperationException("No active catalogue document is available.");
			}
			return dataSet;
		}

		private DataTable ModelsTable()
		{
			return RequireTable("Models");
		}

		private DataTable RequireTable(string name)
		{
			DataTable dataTable = Document().Tables[name];
			if (dataTable == null)
			{
				throw new InvalidOperationException("Catalogue table '" + name + "' is missing.");
			}
			return dataTable;
		}

		private DataRow FindRow(string identifier)
		{
			if (string.IsNullOrWhiteSpace(identifier))
			{
				return null;
			}
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = ModelsTable().Rows.GetEnumerator();
				while (enumerator.MoveNext())
				{
					DataRow dataRow = (DataRow)enumerator.Current;
					if (dataRow.RowState != DataRowState.Deleted && string.Equals(Convert.ToString(RuntimeHelpers.GetObjectValue(dataRow["Identifier"])), identifier, StringComparison.OrdinalIgnoreCase))
					{
						return dataRow;
					}
				}
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
			return null;
		}

		private CassetteModel Map(DataRow row)
		{
			DateTime addedAt = DateTime.MinValue;
			if (!row.IsNull("Date"))
			{
				addedAt = Convert.ToDateTime(RuntimeHelpers.GetObjectValue(row["Date"]));
			}
			return new CassetteModel(ResolveBrandCode(Convert.ToString(RuntimeHelpers.GetObjectValue(row["Brand"]))), ReadInteger(row, "Type"), Convert.ToString(RuntimeHelpers.GetObjectValue(row["Model"])), Convert.ToString(RuntimeHelpers.GetObjectValue(row["Code"])), Convert.ToString(RuntimeHelpers.GetObjectValue(row["Identifier"])), Convert.ToString(RuntimeHelpers.GetObjectValue(row["Name"])), ReadInteger(row, "Number"), addedAt, Convert.ToString(RuntimeHelpers.GetObjectValue(row["Notes"])));
		}

		private string ResolveBrandCode(string storedValue)
		{
			IEnumerator enumerator = default(IEnumerator);
			try
			{
				enumerator = RequireTable("Brands").Rows.GetEnumerator();
				while (enumerator.MoveNext())
				{
					DataRow dataRow = (DataRow)enumerator.Current;
					if (dataRow.RowState != DataRowState.Deleted)
					{
						string text = Convert.ToString(RuntimeHelpers.GetObjectValue(dataRow["Code"]));
						string b = Convert.ToString(RuntimeHelpers.GetObjectValue(dataRow["Brand"]));
						if (!string.Equals(storedValue, text, StringComparison.OrdinalIgnoreCase) && !string.Equals(storedValue, b, StringComparison.OrdinalIgnoreCase))
						{
							continue;
						}
						return text;
					}
				}
			}
			finally
			{
				if (enumerator is IDisposable)
				{
					(enumerator as IDisposable).Dispose();
				}
			}
			return storedValue;
		}

		private static int ReadInteger(DataRow row, string columnName)
		{
			if (row.IsNull(columnName))
			{
				return 0;
			}
			return Convert.ToInt32(RuntimeHelpers.GetObjectValue(row[columnName]));
		}

		private void SynchronizeModelCounter()
		{
			DataRow dataRow = RequireTable("Counters").Rows.Find("Models");
			if (dataRow != null)
			{
				dataRow["Number"] = ModelsTable().Rows.Count;
			}
		}
	}
}
